src/submodules/io/jld2_integration.jl

```diff
@@ -0,0 +1,199 @@
+"""
+to_serializable(t) = t
+
+"""
+    AgentsIO.from_serializable(t; kwargs...)
+
+Given a value in its serializable form, return the original version. This defaults
+to the value itself, unless a more specific method is defined. Define a method for 
+this function and for [`AgentsIO.to_serializable`](@ref) if you need custom
+serialization for model properties. This also enables passing keyword arguments
+to [`AgentsIO.load_checkpoint`](@ref) and having access to them through `kwargs`.
+This function is not called recursively on every type/value during deserialization. The final
+serialization functionality is enabled by JLD2.jl. To define custom serialization
+for every occurence of a specific type (such as agent structs), refer to the
+Custom Serialization section of JLD2.jl documentation.
+"""
+from_serializable(t; kwargs...) = t
+
+struct SerializableABM{S,A<:AbstractAgent,P,R<:AbstractRNG}
+    agents::Vector{A}
+    space::S
+    properties::P
+    rng::R
+    maxid::Int64
+end
+
+struct SerializableGridSpace{D,P,W}
+    dims::NTuple{D,Int}
+    metric::Symbol
+    hoods::Vector{Tuple{Float64,Agents.Hood{D}}}
+    hoods_tuple::Vector{Tuple{NTuple{D,Float64},Agents.Hood{D}}}
+    pathfinder::W
+end
+
+struct SerializableContinuousSpace{D,P,T<:AbstractFloat}
+    grid::SerializableGridSpace{D,P}
+    dims::NTuple{D,Int}
+    spacing::T
+    extent::NTuple{D,T}
+end
+
+struct SerializableGraphSpace{G}
+    graph::G
+end
+
+struct SerializableAStar{D,P,M}
+    agent_paths::Vector{Tuple{Int,Vector{Dims{D}}}}
+    grid_dims::Dims{D}
+    neighborhood::Vector{Dims{D}}
+    admissibility::Float64
+    walkable::BitArray{D}
+    cost_metric::Pathfinding.CostMetric{D}
+end
+
+to_serializable(t::ABM) = SerializableABM(
+    collect(values(t.agents)),
+    to_serializable(t.space),
+    to_serializable(t.properties),
+    t.rng,
+    t.maxid.x,
+)
+
+function to_serializable(t::GridSpace{D,P,W}) where {D,P,W}
+    pathfinder = to_serializable(t.pathfinder)
+    SerializableGridSpace{D,P,typeof(pathfinder)}(
+        size(t.s),
+        t.metric,
+        [(k, v) for (k, v) in t.hoods],
+        [(k, v) for (k, v) in t.hoods_tuple],
+        pathfinder,
+    )
+end
+
+to_serializable(t::ContinuousSpace{D,P,T}) where {D,P,T} =
+    SerializableContinuousSpace{D,P,T}(
+        to_serializable(t.grid),
+        t.dims,
+        t.spacing,
+        t.extent,
+    )
+
+to_serializable(t::GraphSpace{G}) where {G} = SerializableGraphSpace{G}(t.graph)
+
+to_serializable(t::Pathfinding.AStar{D,P,M}) where {D,P,M} =
+    SerializableAStar{D,P,M}(
+        [(k, collect(v)) for (k, v) in t.agent_paths],
+        t.grid_dims,
+        map(Tuple, t.neighborhood),
+        t.admissibility,
+        t.walkable,
+        t.cost_metric,
+    )
+
+function from_serializable(t::SerializableABM{S,A}; kwargs...) where {S,A}
+    abm = ABM(
+        A,
+        from_serializable(t.space; kwargs...);
+        scheduler = get(kwargs, :scheduler, Schedulers.fastest),
```

Comment:
In terms of speed, `get` is surprisingly slow. We use [haskey](https://github.com/JuliaDynamics/Agents.jl/blob/cc623023755e77ca49e72e5a97156ea342b48be5/src/spaces/grid.jl#L168-L172) throughout the space generation routines because of this.

Unsure if this is truly a problem in these routines, but it may be worth checking.

Reply:
[`kwargs` isn't a `NamedTuple`](https://stackoverflow.com/questions/59296403/how-do-i-retrieve-keyword-arguments-out-of-a-field-of-splatted-kwargs), so (as far as I'm aware) accessing it's fields requires converting it to `NamedTuple`/`Dict`/similar or using `get`. I'll compare speeds of the different methods to see which one works out faster, and if there's much of a difference at all.

Reply:
Ah, sure. Yeah, I'm talking about `get` from `Dict`, so my point is moot.

Reply:
In my tests, both `get` and the entire `haskey` structure took the same time